### PR TITLE
Update dnn.cpp

### PR DESCRIPTION
**Pull request by MarkGHX**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/dnn.cpp [hunk 1]
```diff
@@ -1120,6 +1126,15 @@ static Ptr<BackendWrapper> wrapMat(int backendId, int targetId, cv::Mat& m)
         return Ptr<BackendWrapper>(new NgraphBackendWrapper(targetId, m));
 #else
         CV_Error(Error::StsNotImplemented, "This OpenCV version is built without support of Inference Engine + nGraph");
+#endif
+    }
+    else if (backendId == DNN_BACKEND_WEBNN)
+    {
+        CV_Assert(haveWebNN());
+#ifdef HAVE_WEBNN
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp:1133]: should `CV_Assert(haveWebNN());` be guarded by `HAVE_WEBNN`?

File: modules/dnn/src/dnn.cpp [hunk 2]
```diff
@@ -1263,6 +1278,12 @@ struct Net::Impl : public detail::NetImplBase
             {
                 return wrapMat(preferableBackend, preferableTarget, host);
             }
+            else if (preferableBackend == DNN_BACKEND_WEBNN)
+            {
+                #ifdef HAVE_WEBNN
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp]: Please keep the indent aligned to others.

File: modules/dnn/src/dnn.cpp [hunk 3]
```diff
@@ -1403,6 +1424,10 @@ struct Net::Impl : public detail::NetImplBase
                   preferableTarget == DNN_TARGET_FPGA
             );
         }
+#endif
+#ifdef HAVE_WEBNN
+        CV_Assert(preferableBackend != DNN_BACKEND_WEBNN ||
+                  preferableTarget == DNN_TARGET_WEBNN);
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp]: Should we assert like
```c++
if (preferableBackend == DNN_BACKEND_WEBNN) {
  CV_Assert(preferableTarget == DNN_TARGET_CPU || preferableTarget == DNN_TARGET_GPU)
}
```

Comment (huningxin) [reply modules/dnn/src/dnn.cpp]: Please define `DNN_BACKEND_WEBNN` and `DNN_TARGET_GPU` in dnn.hpp.

File: modules/dnn/src/dnn.cpp [hunk 4]
```diff
@@ -1626,6 +1651,14 @@ struct Net::Impl : public detail::NetImplBase
             initNgraphBackend(blobsToKeep_);
 #else
             CV_Error(Error::StsNotImplemented, "This OpenCV version is built without support of Inference Engine + nGraph");
+#endif
+        }
+        else if (preferableBackend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp]: Please change the condition to webnn.

File: modules/dnn/src/dnn.cpp [hunk 5]
```diff
@@ -2332,6 +2365,13 @@ struct Net::Impl : public detail::NetImplBase
     }
 #endif  // HAVE_DNN_NGRAPH
 
+    void initNgraphBackend(const std::vector<LayerPin>& blobsToKeep_)
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp]: `initNgraphBackend` seems not correct.

File: modules/dnn/src/dnn.cpp [hunk 6]
```diff
@@ -3368,6 +3408,10 @@ struct Net::Impl : public detail::NetImplBase
                     forwardInfEngine(ld.outputBlobsWrappers, node, isAsync);
                 }
                 else if (preferableBackend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
+                {
+                    forwardNgraph(ld.outputBlobsWrappers, node, isAsync);
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp:3415]: ??

File: modules/dnn/src/dnn.cpp [hunk 7]
```diff
@@ -3368,6 +3408,10 @@ struct Net::Impl : public detail::NetImplBase
...
+                }
+                 else if (preferableBackend == DNN_BACKEND_WEBNN)
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp:3417]: please fix this condition

File: modules/dnn/src/dnn.cpp [hunk 8]
```diff
@@ -1626,6 +1651,14 @@ struct Net::Impl : public detail::NetImplBase
             initNgraphBackend(blobsToKeep_);
 #else
             CV_Error(Error::StsNotImplemented, "This OpenCV version is built without support of Inference Engine + nGraph");
+#endif
+        }
+        else if (preferableBackend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
+        {
+#ifdef HAVE_WEBNN
+            initWebNNBackend(blobsToKeep_);
```
Comment (huningxin) [inline modules/dnn/src/dnn.cpp:1662]: Please put a template implementation of initWebNNBackend in this file.

**Review by huningxin — CHANGES_REQUESTED** [review]
Please make this PR to pass the build.

File: modules/dnn/src/dnn.cpp (hunk 1 above)
Comment (MarkGHX) [reply modules/dnn/src/dnn.cpp:1133]: fixed.

File: modules/dnn/src/dnn.cpp (hunk 2 above)
Comment (MarkGHX) [reply modules/dnn/src/dnn.cpp]: fixed

File: modules/dnn/src/dnn.cpp (hunk 3 above)
Comment (MarkGHX) [reply modules/dnn/src/dnn.cpp]: fixed

File: modules/dnn/src/op_webnn.cpp [hunk 9]
```diff
@@ -0,0 +1,7 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018-2019, Intel Corporation, all rights reserved.
+// Third party copyrights are property of their respective owners.
+
```
Comment (huningxin) [inline modules/dnn/src/op_webnn.cpp]: Please remove this file from this PR. You can add it in your next PR.

File: modules/dnn/src/op_webnn.hpp [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+    virtual bool supportBackend(int backendId) CV_OVERRIDE;
+
+private:
+    InferenceEngine::CNNNetwork t_net;
+};
+
+#endif  // HAVE_WebNN
+
+constexpr bool haveWebNN() {
```
Comment (huningxin) [inline modules/dnn/src/op_webnn.hpp:26]: Please only keep this function and remove others. Please add others in your next PR.

File: modules/dnn/src/op_webnn.hpp [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+#endif
+    }
+
+void forwardWebNN(const std::vector<Ptr<BackendWrapper> >& outBlobsWrappers,
+                   Ptr<BackendNode>& node, bool isAsync);
+
+}}  // namespace cv::dnn
+
+
+#endif  // __OPENCV_DNN_IE_NGRAPH_HPP__
```
Comment (huningxin) [inline modules/dnn/src/op_webnn.hpp]: `__OPENCV_DNN_OP_WEBNN_HPP__`

**Review by huningxin — APPROVED** [review]

**Pull request merged**